File: services/archivist/es-similarity/src/main/java/com/zorroa/es/similarity/SimilarityPlugin.java
package com.zorroa.es.similarity;

import org.apache.lucene.index.LeafReaderContext;
import org.apache.lucene.util.BytesRef;
import org.elasticsearch.common.settings.Settings;
import org.elasticsearch.index.fielddata.ScriptDocValues;
import org.elasticsearch.plugins.Plugin;
import org.elasticsearch.plugins.ScriptPlugin;
import org.elasticsearch.script.ScoreScript;
import org.elasticsearch.script.ScriptContext;
import org.elasticsearch.script.ScriptEngine;
import org.elasticsearch.script.SearchScript;

import java.io.IOException;
import java.util.*;
import java.util.logging.Logger;
import java.util.stream.Collectors;

public class SimilarityPlugin extends Plugin implements ScriptPlugin {

    private static final double NO_SCORE = 0;

    private static final Logger logger = Logger.getLogger(SimilarityPlugin.class.getCanonicalName());

    @Override
    public ScriptEngine getScriptEngine(Settings settings, Collection<ScriptContext<?>> ctx) {
        return new SimilarityEngine();
    }

    private static class SimilarityEngine implements ScriptEngine {

        @Override
        public String getType() {
            return "zorroa-similarity";
        }

        @Override
        public <FactoryType> FactoryType compile(String name, String code, ScriptContext<FactoryType> context, Map<String, String> params) {

            if ("similarity".equals(code)) {
                ScoreScript.Factory factory = (p, lookup) -> new ScoreScript.LeafFactory() {

                    private static final double NORM = 100.0;

                    private String field;
                    private final List<String> charHashes;
                    private final List<Float> weights;
                    private int length = 0;
                    private double minScore = 0;
                    private int resolution;
                    private final boolean header;
                    private final char version;
                    private final int dataPos;
                    private final int numHashes;
                    private final double possibleScore;
                    private final double singleScore;

                    {
                        field = p.get("field").toString();
                        minScore =  Double.valueOf(p.getOrDefault("minScore", "1").toString()) / NORM;
                        resolution = 15;

                        List<String> _hashes = Arrays.asList(p.get("hashes").toString().split(","));
                        List<Float> _weights = Arrays.asList(p.get("weights").toString().split(",")).stream()
                                .map(e-> Float.valueOf(e.toString())).collect(Collectors.toList());

                        if (_hashes == null) {
                            throw new IllegalArgumentException(
                                    "Hashes cannot be null");
                        }

                        if (_weights == null) {
                            _weights = Collections.nCopies(_hashes.size(), 1.0f);
                        }

                        if (_hashes.size() != _weights.size()) {
                            throw new IllegalArgumentException(
                                    "HammingDistanceScript weights must align with hashes");
                        }

                        /**
                         * Go through all the values and remove the null
                         * values and populate the charHashes and
                         * weights fields with valid values.
                         */
                        charHashes = new ArrayList();
                        weights = new ArrayList();
                        for (int i=0; i<_hashes.size(); i++) {
                            String hash = _hashes.get(i);
                            if (hash == null || hash.isEmpty()) {
                                continue;
                            }
                            charHashes.add(hash);
                            weights.add(_weights.get(i));
                        }

                        /**
                         * If there are no valid hashes left, initialize to defaults
                         */
                        if (charHashes.isEmpty()) {
                            singleScore = possibleScore = numHashes = dataPos = version = 0;
                            header = false;
                        }
                        else {
                            /**
                             * Use the first hash to determine if there is a header.
                             */
                            String hash = charHashes.get(0);
                            header = hash.charAt(0) == '#';
                            length = hash.length();
                            numHashes = charHashes.size();

                            /**
                             * TODO: more sophisticated header parsing.
                             *
                             * There are 2 fields every has leads with:
                             * 1 char: version
                             * 2 chars: position of data (called "headerSize" here)
                             *
                             * A version 0 hash has 1 field, resolution.
                             */
                            if (header) {
                                version = hash.charAt(1);

                                // The start position of the data.
                                dataPos = Integer.parseInt(hash.substring(2, 4), 16);

                                if (version <= 0) {
                                    // Resolution is the next byte.
                                    resolution = Integer.parseInt(hash.substring(4, 6), 16);
                                }
                            } else {
                                version = 0;
                                dataPos = 0;
                            }

                            // To get the proper score, we subtract header size from the length here.
                            singleScore = resolution * (length - dataPos);
                            possibleScore = singleScore * numHashes;
                        }
                    }

                    @Override
                    public ScoreScript newInstance(LeafReaderContext ctx) throws IOException {
                        return new ScoreScript(p, lookup, ctx) {

                            @Override
                            public double execute() {

                                ScriptDocValues.Strings strings;

                                if (getDoc().containsKey(field)) {
                                    strings = (ScriptDocValues.Strings) getDoc().get(field);
                                }
                                else {
                                    return NO_SCORE;
                                }

                                double score = charHashesComparison(strings.getValue());
                                return score >= minScore ? score : NO_SCORE;
                            }

                            public final double charHashesComparison(String fieldValue) {
                                double score = 0;
                                if (possibleScore == 0) {
                                    return NO_SCORE;
                                }

                                if (fieldValue == null || fieldValue.length() == 0) {
                                    return NO_SCORE;
                                }

                                char[] bytes = fieldValue.toCharArray();
                                char ver = bytes[1];
                                for (int i = 0; i < numHashes; ++i) {
                                    String hash = charHashes.get(i);
                                    if (hash == null) {
                                        continue;
                                    }
                                    if (header) {
                                        if (ver != hash.charAt(1)) {
                                            continue;
                                        }
                                    }
                                    else {
                                        if (bytes.length != hash.length()) {
                                            continue;
                                        }
                                    }
                                    score += (weights.get(i) * hammingDistance(bytes, hash));
                                }
                                score = normalize(score);
                                return score;
                            }

                            public final double normalize(double score) {
                                score = (score / possibleScore);
                                return score;
                            }

                            public final double hammingDistance(final char[] lhs, final String rhs) {
                                double score = singleScore;
                                for (int i = dataPos; i < length; i++) {
                                    score -= Math.abs(lhs[i] - rhs.charAt(i));
                                }
                                return score;
                            }

                            public int getResolution() {
                                return resolution;
                            }
                            public String getField() { return field; }
                            public int getNumHashes() {
                                return numHashes;
                            }
                        };
                    }

                    @Override
                    public boolean needs_score() {
                        return false;
                    }
                };

                return context.factoryClazz.cast(factory);
            }

            throw new IllegalArgumentException("Unknown script name " + code);
        }
    }

}

File: services/elasticsearch/es-similarity/src/main/kotlin/com/zorroa/zmlp/similarity/SimilarityPlugin.kt
package com.zorroa.zmlp.similarity

import org.apache.lucene.index.LeafReaderContext
import org.elasticsearch.common.settings.Settings
import org.elasticsearch.index.fielddata.ScriptDocValues
import org.elasticsearch.plugins.Plugin
import org.elasticsearch.plugins.ScriptPlugin
import org.elasticsearch.script.ScoreScript
import org.elasticsearch.script.ScriptContext
import org.elasticsearch.script.ScriptEngine
import org.elasticsearch.search.lookup.SearchLookup
import java.util.logging.Logger
import kotlin.math.abs

class SimilarityPlugin : Plugin(), ScriptPlugin {

    override fun getScriptEngine(settings: Settings?, ctx: Collection<ScriptContext<*>>?): ScriptEngine {
        return SimilarityEngine()
    }

    private class SimilarityFactory : ScoreScript.Factory {
        override fun newFactory(params: MutableMap<String, Any>, lookup: SearchLookup): ScoreScript.LeafFactory {
            return SimilarityLeafFactory(params, lookup)
        }
    }

    private class SimilarityEngine : ScriptEngine {

        override fun getSupportedContexts(): Set<ScriptContext<*>> {
            return setOf(ScoreScript.CONTEXT)
        }

        override fun getType(): String {
            return "zorroa-similarity"
        }

        override fun <T> compile(
            name: String,
            code: String,
            context: ScriptContext<T>,
            params: Map<String, String>
        ): T {

            if ("similarity" == code) {
                return context.factoryClazz.cast(SimilarityFactory()) as T
            }

            throw IllegalArgumentException("Unknown script name $code")
        }
    }

    private class SimilarityLeafFactory constructor(
        val params: MutableMap<String, Any>,
        val lookup: SearchLookup
    ) : ScoreScript.LeafFactory {

        private val field: String = params.getValue("field") as String
        private val charHashes: List<String>
        private val length: Int
        private val minScore: Double = params.getOrDefault("minScore", 0.75) as Double
        private val resolution: Int = 16
        private val numHashes: Int
        private val singleScore: Double

        init {

            if (params["hashes"] == null) {
                throw IllegalArgumentException("Hashes cannot be null")
            }

            charHashes = (params["hashes"] as List<String>)
                .mapNotNull { it }
                .filter { it.isNotEmpty() }

            /**
             * If there are no valid hashes left, initialize to defaults
             */
            if (charHashes.isEmpty()) {
                singleScore = 0.0
                numHashes = 0
                length = 0
            } else {
                val hash = charHashes[0]
                length = hash.length
                numHashes = charHashes.size
                singleScore = (resolution * length).toDouble()
            }
        }

        override fun needs_score(): Boolean {
            return true
        }

        override fun newInstance(ctx: LeafReaderContext?): ScoreScript {

            return object : ScoreScript(params, lookup, ctx) {

                override fun execute(explanationHolder: ExplanationHolder?): Double {

                    val strings: ScriptDocValues.Strings

                    if (doc[field]?.size ?: 0 > 0) {
                        strings = doc[field] as ScriptDocValues.Strings
                    } else {
                        return noScore
                    }

                    val score = charHashesComparison(strings.value)
                    return if (score >= minScore) score else noScore
                }

                fun charHashesComparison(fieldValue: String?): Double {
                    var totalScore = 0.0

                    if (numHashes == 0) {
                        return noScore
                    }

                    if (fieldValue == null || fieldValue.isEmpty()) {
                        return noScore
                    }

                    val bytes = fieldValue.toCharArray()
                    for (i in 0 until numHashes) {
                        val hash = charHashes[i]
                        if (bytes.size != hash.length) {
                            continue
                        }
                        val elementScore = normalize(hammingDistance(bytes, hash))
                        if (elementScore < minScore) {
                            return noScore
                        } else {
                            totalScore += elementScore
                        }
                    }

                    return totalScore
                }

                fun normalize(score: Double): Double {
                    return score / singleScore
                }

                fun hammingDistance(lhs: CharArray, rhs: String): Double {
                    var score = singleScore
                    for (i in 0 until length) {
                        score -= abs(lhs[i] - rhs[i])
                    }
                    return score
                }
            }
        }
    }

    companion object {

        private const val normFactor = 100.0

        private const val noScore = 0.0

        private val logger = Logger.getLogger(SimilarityPlugin::class.java.canonicalName)
    }
}
